File: RealLife.TypeScript/script/ScriptBase.ts
/// <reference path="../Objeto.ts"/>

module RealLife
{
    // #region Importações
    // #endregion Importações

    // #region Enumerados
    // #endregion Enumerados

    export abstract class ScriptBase extends Objeto
    {
        // #region Constantes
        // #endregion Constantes

        // #region Atributos

        private _arrObjEtapa: Array<EtapaBase>;

        private get arrObjEtapa(): Array<EtapaBase>
        {
            if (this._arrObjEtapa != null)
            {
                return this._arrObjEtapa;
            }

            this._arrObjEtapa = new Array<EtapaBase>();

            return this._arrObjEtapa;
        }

        // #endregion Atributos

        // #region Construtores
        // #endregion Construtores

        // #region Métodos

        public addEtapa(objEtapa: EtapaBase): void
        {
            if (objEtapa == null)
            {
                return;
            }

            if (this.arrObjEtapa.indexOf(objEtapa) > -1)
            {
                return;
            }

            this.arrObjEtapa.push(objEtapa);
        }

        public destruir(): void
        {
            super.destruir();

            for (var i = 0; i < this.arrObjEtapa.length; i++)
            {
                this.arrObjEtapa[i].destruir();
            }
        }

        // #endregion Métodos

        // #region Eventos
        // #endregion Eventos
    }
}

File: ModTool/ScriptBase.cs
using GTA;

namespace ModTool
{
    public abstract class ScriptBase : Script
    {
        #region Constantes

        #endregion Constantes

        #region Atributos

        #endregion Atributos

        #region Construtores

        public ScriptBase()
        {
            this.iniciar();
        }

        #endregion Construtores

        #region Métodos

        public void notificar(string strNotificacao, params object[] arrObjParam)
        {
            if (string.IsNullOrEmpty(strNotificacao))
            {
                return;
            }

            if (arrObjParam != null)
            {
                strNotificacao = string.Format(strNotificacao, arrObjParam);
            }

            UI.Notify(strNotificacao);
        }

        protected virtual void inicializar()
        {
        }

        protected virtual void setEventos()
        {
        }

        private void iniciar()
        {
            this.inicializar();
            this.setEventos();
        }

        #endregion Métodos

        #region Eventos

        #endregion Eventos
    }
}